text/0000-bang-type.md

```diff
@@ -0,0 +1,401 @@
+never contain an element, a `!` enum variant makes the variant guaranteed never
+to occur and so forth. It might seem pointless to use a `!` function argument
+or a `Vec<!>` (just as it would be pointless to use a `()` function argument or
+a `Vec<()>`), but that's no reason to disallow it. And generic code sometimes
+requires it.
+
+Rust already has empty types in the form of empty enums. Any code that could be
+written with this RFC's `!` can already be written by swapping out `!` with
+`Never` (sans implicit casts, see below). So if this RFC could create any
+issues for the language (such as making it unsound or complicating the
+compiler) then these issues would already exist for `Never`.
+
+# Detailed design
+
+Add a type `!` to Rust. `!` behaves like an empty enum except that it can be
+implicitly cast to any other type. ie. the following code is acceptable:
+
+```rust
+let r: Result<i32, !> = Ok(23);
+let i = match r {
+    Ok(i)   => i,
+    Err(e)  => e, // e is cast to i32
+}
+```
+
+Implicit casting is necessary for backwards-compatibility so that code like the
+following will continue to compile:
+
+```rust
+let i: i32 = match some_bool {
+    true  => 23,
+    false => panic!("aaah!"), // an expression of type `!`, gets cast to `i32`
+}
+
+match break {
+    ()  => 23,  // matching with a `()` forces the match argument to be cast to type `()`
+}
+```
+
+In the compiler, remove the distinctions that treat diverging and converging
+expressions as two different kinds of things (eg. stuff like `FnConverging` vs
+`FnDiverging`). Use the type system to do things like reachability analysis.
+
+Add an implementation for `!` of any trait that it can trivially implement. Add
+methods to `Result<T, !>` and `Result<!, E>` for safely extracting the inner
+value. Name these methods along the lines of `unwrap_nopanic`, `safe_unwrap` or
+something.
+
+# Drawbacks
+
+Someone would have to implement this.
+
+# Alternatives
+
+  * Don't do this.
+  * Move @reem's `Void` type into `libcore`. This would create a standard empty
+    type and make it available for use in the standard libraries. If we were to
+    do this it might be an idea to rename `Void` to something else (`Never`,
+    `Empty` and `Mu` have all been suggested). Although `Void` has some
+    precedence in languages like Haskell and Idris the name is likely to trip
+    up people coming from a C/Java et al. background as `Void` is *not* `void`
+    but it can be easy to confuse the two.
+
+# Unresolved questions
+
+Apparently, rust used to have something similar to this but it was removed.
+There are still a few references to `ty_bot` in the compiler. Why was this
+taken out? Note that if there any arguments for not having type `!` in the
+language they should apply equally well to `Never`/`Void` so I assume the old
+`ty_bot` was trying to be something crazier than this RFC's `!` (such as a
+subtype of all types, given the name). Could someone who was around back then
+clarify this?
```

Comment:
> Do you want `Option<!>` to be usable as an `Option<T>`? Or `fn() -> !` as `fn() -> T`?
If not, the change is backwards-incompatible.

No, I don't think `!` should be a subtype of `T`. Can you give me an example of code that this would break? `Option<!>` isn't currently a type so I don't see how that's an issue. And `fn() -> !` isn't currently a subtype of `fn() -> T` either. The kind of silent coercion I'm thinking of is if you have

```rust
match b {
    true  => 0i32,       // type i32
    false => abort(0),   // type `!`
}
```

The typechecker will see the mismatch and change it to.

```rust
match b {
    true  => 0i32,
    false => abort(0) as i32,
}
```

Nothing deeper than that. That should be enough to continue accepting code that uses a diverging expression where another type is expected (am I wrong?)

> And I mean "what is the value in accepting invalid code which is unreachable?".
If you wrote `panic!() + 0`, you can just delete the `+ 0`. I really see no point in this.

I don't think we should accept invalid code which is unreachable. I mean, I suppose we *could*, but I'd be uneasy about accepting any code whatsoever just because we know it's unreachable. If we did, should that behaviour extend to all empty types or just `!`? We don't currently accept

```rust
let x: Void = panic!("oh no!");
let y = x + 0;
```